include/unifex/task.hpp

```diff
@@ -180,76 +181,90 @@ struct _task<T>::type {
     std::optional<continuation_info> info_;
     inplace_stop_token stoken_;
   };
+};
 
-  template<
-    template<typename...> class Variant,
-    template<typename...> class Tuple>
-  using value_types = Variant<
-    typename std::conditional_t<
-      std::is_void_v<T>, type_list<>, type_list<T>>
-    ::template apply<Tuple>>;
-
-  template<
-    template<typename...> class Variant>
-  using error_types = Variant<std::exception_ptr>;
-
-  static constexpr bool sends_done = true;
-
-  explicit type(coro::coroutine_handle<promise_type> h) noexcept
-      : coro_(h) {}
-
-  ~type() {
-    if (coro_)
-      coro_.destroy();
-  }
-
-  type(type&& t) noexcept : coro_(std::exchange(t.coro_, {})) {}
-
-  type& operator=(type t) noexcept {
-    std::swap(coro_, t.coro_);
-    return *this;
-  }
+template<typename ThisPromise, typename OtherPromise>
+struct _awaiter {
+  struct type {
+    using result_type = typename ThisPromise::result_type;
 
-private:
-  template<typename Promise>
-  struct awaiter {
-    explicit awaiter(coro::coroutine_handle<promise_type> coro) noexcept
+    explicit type(coro::coroutine_handle<ThisPromise> coro) noexcept
     : coro_(coro)
     {}
 
-    awaiter(awaiter&& other) noexcept
+    type(type&& other) noexcept
     : coro_(std::exchange(other.coro_, {}))
     {}
 
-    ~awaiter() {
+    ~type() {
       if (coro_) coro_.destroy();
     }
 
     bool await_ready() noexcept {
       return false;
     }
 
-    coro::coroutine_handle<promise_type> await_suspend(
-        coro::coroutine_handle<Promise> h) noexcept {
+    coro::coroutine_handle<ThisPromise> await_suspend(
+        coro::coroutine_handle<OtherPromise> h) noexcept {
       assert(coro_);
       coro_.promise().continuation_ = h;
-      coro_.promise().doneCallback_ = &forward_unhandled_done_callback<Promise>;
+      coro_.promise().doneCallback_ = &forward_unhandled_done_callback<OtherPromise>;
       coro_.promise().info_.emplace(
           continuation_info::from_continuation(h.promise()));
```

Comment:
I wonder how much cost the support for `visit_continuations` here has in terms of binary-size?

Reply:
You could put it in a branch and ask Ian to try it out.